### PR TITLE
Implement `BigDec` binary search in osmoutils

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Progress towards: #2794 and #2702

## What is the purpose of the change

This PR implements a `BigDec` binary search function for our general osmoutils package.

## Brief Changelog

- Add `BinarySearchBigDec` to osmoutils
- Add `CompareBigDec` to osmoutils
- Refactor existing and new tests to be more readable (string map names, `t.Run` etc.)

## Testing and Verifying

This change added tests for both new functions in `osmoutils/binary_search_test.go`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: osmoutils/binary_search.go [hunk 1]
```diff
@@ -93,3 +133,41 @@ func BinarySearch(f func(input sdk.Int) (sdk.Int, error),
...
+	}
+	curIteration := 0
+	for ; curIteration < maxIterations; curIteration += 1 {
+		compRes := errTolerance.CompareBigDec(curOutput, targetOutput)
+		if compRes > 0 {
+			upperbound = curEstimate
+		} else if compRes < 0 {
+			lowerbound = curEstimate
+		} else {
+			break
```
Comment (p0mvn) [inline osmoutils/binary_search.go:161]: Consider simply returning here

File: osmoutils/binary_search.go [hunk 2]
```diff
@@ -93,3 +133,41 @@ func BinarySearch(f func(input sdk.Int) (sdk.Int, error),
...
+		curEstimate = lowerbound.Add(upperbound).Quo(osmomath.NewBigDec(2))
+		curOutput, err = f(curEstimate)
+		if err != nil {
+			return osmomath.BigDec{}, err
+		}
+	}
+	if curIteration == maxIterations {
+		return osmomath.BigDec{}, errors.New("hit maximum iterations, did not converge fast enough")
+	}
+	return curEstimate, nil
```
Comment (p0mvn) [inline osmoutils/binary_search.go:172]: If you go with the previous suggestion, you should be able to simply return error here, without doing the iteration check

File: osmoutils/binary_search.go (hunk 1 above)
Comment (AlpinYukseloglu) [reply osmoutils/binary_search.go]: This is a good point – after a bit of thought this does seem to be the simplest approach. Updated!

File: osmoutils/binary_search.go (hunk 2 above)
Comment (AlpinYukseloglu) [reply osmoutils/binary_search.go]: Updated, please see my comment above! 

https://github.com/osmosis-labs/osmosis/pull/2802/files#r975799847

File: osmoutils/binary_search.go [hunk 3]
```diff
@@ -56,6 +58,44 @@ func (e ErrTolerance) Compare(expected sdk.Int, actual sdk.Int) int {
 	return 0
 }
 
+// CompareBigDec returns if actual is within errTolerance of expected.
+// returns 0 if it is
+// returns 1 if not, and expected > actual.
+// returns -1 if not, and expected < actual
```
Comment (p0mvn) [inline osmoutils/binary_search.go:64]:
```suggestion
// CompareBigDec validates if actual is within errTolerance of expected.
// returns 0 if it is
// returns 1 if not, and expected > actual.
// returns -1 if not, and expected < actual
```

File: osmoutils/binary_search_test.go [hunk 4]
```diff
@@ -41,26 +43,89 @@ func TestBinarySearch(t *testing.T) {
...
+		"exponential f, large additive err tolerance, does not converge": {expF, osmomath.ZeroDec(), osmomath.NewBigDec(1 << 50), osmomath.NewBigDec((1 << 30)), testErrToleranceAdditive, 10, osmomath.BigDec{}, true},
+		"exponential f, large multiplicative err tolerance, converges": {expF, osmomath.ZeroDec(), osmomath.NewBigDec(1 << 50), osmomath.NewBigDec(1 + (1 << 25)), testErrToleranceMultiplicative, 51, osmomath.NewBigDec(322539792367616), false},
+		"exponential f, large multiplicative err tolerance, does not converge": {expF, osmomath.ZeroDec(), osmomath.NewBigDec(1 << 50), osmomath.NewBigDec(1 + (1 << 25)), testErrToleranceMultiplicative, 10, osmomath.BigDec{}, true},
+		"exponential f, both err tolerances, converges": {expF, osmomath.ZeroDec(), osmomath.NewBigDec(1 << 50), osmomath.NewBigDec((1 << 15)), testErrToleranceBoth, 51, osmomath.NewBigDec(1 << 45), false},
+		"exponential f, both err tolerances, does not converge": {expF, osmomath.ZeroDec(), osmomath.NewBigDec(1 << 50), osmomath.NewBigDec((1 << 30)), testErrToleranceBoth, 10, osmomath.BigDec{}, true},
```
Comment (p0mvn) [inline osmoutils/binary_search_test.go:116]: Consider test cases with different lower and upper bounds. Happy for this to be done in another issue / PR.

**Review by p0mvn — APPROVED** [review]
LGTM. What are our plans for the binary search solver in `stableswap`? Is it going to be used anywhere?

**Comment by AlpinYukseloglu** [comment]
Yes! After we merge this and #2697 (which has most of our stableswap CFMM tests in it) I will migrate our core solvers to use this implementation @p0mvn

File: osmoutils/binary_search_test.go (hunk 4 above)
Comment (AlpinYukseloglu) [reply osmoutils/binary_search_test.go:116]: Created issue here! #2807

File: osmoutils/binary_search.go [hunk 5]
```diff
@@ -56,6 +58,44 @@ func (e ErrTolerance) Compare(expected sdk.Int, actual sdk.Int) int {
...
+	// Check additive tolerance equations
+	if !e.AdditiveTolerance.IsNil() {
+		// if no error accepted, do a direct compare.
+		if e.AdditiveTolerance.IsZero() {
+			if expected.Equal(actual) {
+				return 0
+			}
+		}
+
+		if diff.GT(osmomath.NewBigDec(e.AdditiveTolerance.Int64())) {
```
Comment (ValarDragon) [inline osmoutils/binary_search.go:84]: I don't think this quite makes sense, to increase the precision, we first truncate to int64 and then increase the size.

Should probably e.AdditiveTolerance.ToDec() and then make a big dec from that.

File: osmoutils/binary_search.go [hunk 6]
```diff
@@ -80,16 +120,50 @@ func BinarySearch(f func(input sdk.Int) (sdk.Int, error),
...
 	}
-	if curIteration == maxIterations {
-		return sdk.Int{}, errors.New("hit maximum iterations, did not converge fast enough")
+
+	return sdk.Int{}, errors.New("hit maximum iterations, did not converge fast enough")
+}
+
+// Binary search BigDec inputs between [lowerbound, upperbound] to a monotonic increasing function f
+// We stop once f(found_input) meets the ErrTolerance constraints.
+// If we perform more than maxIterations (or equivalently lowerbound = upperbound), we return an error.
```
Comment (ValarDragon) [inline osmoutils/binary_search.go:137]:
```suggestion
// BinarySearchBigDec takes as input:
// * an input range [lowerbound, upperbound] 
// * an increasing function f
// * a target output x
// * max number of iterations (for gas control / handling does-not-converge cases)
// 
// It binary searches on the input range, until it finds an input y s.t. f(y) meets the err tolerance constraints for how close it is to x.
// If we perform more than maxIterations (or equivalently lowerbound = upperbound), we return an error.
```

File: osmoutils/binary_search_test.go [hunk 7]
```diff
@@ -41,26 +43,89 @@ func TestBinarySearch(t *testing.T) {
...
+	}
+}
+
+func TestBinarySearchBigDec(t *testing.T) {
+	// straight line function that returns input. Simplest to binary search on,
+	// binary search directly reveals one bit of the answer in each iteration with this function.
+	lineF := func(a osmomath.BigDec) (osmomath.BigDec, error) {
+		return a, nil
+	}
+	expF := func(a osmomath.BigDec) (osmomath.BigDec, error) {
```
Comment (ValarDragon) [inline osmoutils/binary_search_test.go:77]: Shouldn't be called `expF`, this is a cubic function

**Review by ValarDragon — APPROVED** [review]

File: osmoutils/binary_search.go (hunk 5 above)
Comment (AlpinYukseloglu) [reply osmoutils/binary_search.go:84]: Ah so we didn't have the `BigDecFromSDKDec` implemented in `BigDec` when I made this PR so what you suggested wasn't possible then, but I actually agree with this change! Updating now

File: osmoutils/binary_search_test.go (hunk 7 above)
Comment (AlpinYukseloglu) [reply osmoutils/binary_search_test.go:77]: Great point, I was extending the variable names used in the integer binary search but you're completely right – updating!

**Pull request merged**